solidity/contracts/AcreMultiAssetVault.sol

```diff
@@ -0,0 +1,279 @@
+    /// @param amount The amount of assets to deposit.
+    /// @param depositOwner The address of the account for which the deposit is
+    ///        being made.
+    /// @return The ID of the deposit.
+    function depositFor(
+        address asset,
+        uint256 amount,
+        address depositOwner
+    ) public returns (uint256) {
+        if (!supportedAssets[asset]) {
+            revert AssetNotSupported();
+        }
+        if (amount == 0) {
+            revert InvalidAmount(amount);
+        }
+        if (depositOwner == address(0) || depositOwner == asset) {
+            revert InvalidDepositOwner(depositOwner);
+        }
+
+        // Generate a new deposit ID.
+        uint256 depositId = ++depositCount;
+
+        emit DepositCreated(depositOwner, asset, depositId, amount);
+
+        // Transfer the asset from the deposit owner to the vault and approve
+        // Mezo Portal to pull them.
+        IERC20(asset).safeTransferFrom(msg.sender, address(this), amount);
+        IERC20(asset).forceApprove(address(mezoPortal), amount);
+
+        // Create a deposit in Mezo Portal.
+        // 0 denotes no lock period for this deposit.
+        mezoPortal.deposit(address(asset), uint96(amount), 0);
+
+        // MezoPortal doesn't return depositId, so we have to read depositCounter
+        // which assigns depositId to the current deposit.
+        uint256 mezoDepositId = mezoPortal.depositCount();
+
+        // Register the deposit info.
+        // slither-disable-next-line reentrancy-benign
+        deposits[depositOwner][asset][depositId] = DepositInfo({
+            balance: amount,
+            mezoDepositId: mezoDepositId
+        });
+
+        return depositId;
+    }
+
+    /// @notice Withdraws all deposited assets from the given deposit.
+    /// @dev It can be called by the deposit owner only.
+    /// @param asset Address of the asset to be withdrawn.
+    /// @param depositId ID of the deposit.
+    /// @param receiver Address of the account to which the assets will be sent.
+    /// @return The amount of assets withdrawn.
+    function withdraw(
+        address asset,
+        uint256 depositId,
+        address receiver
+    ) external returns (uint256) {
+        if (receiver == address(0) || receiver == asset) {
+            revert InvalidReceiver(receiver);
+        }
+
+        uint256 depositedAmount = deposits[msg.sender][asset][depositId]
+            .balance;
+
+        // Check if the deposit exists. If the deposit was not created or was
+        // already withdrawn, the depositedAmount will be 0.
+        if (depositedAmount == 0) {
+            revert DepositNotFound();
+        }
+
+        // Withdraw the assets from the MezoPortal. Check if the withdrawn amount
+        // matches the deposited amount.
+        uint256 initialBalance = IERC20(asset).balanceOf(address(this));
+
+        // slither-disable-next-line reentrancy-no-eth
+        mezoPortal.withdraw(
+            address(asset),
```

Comment:
This casting looks redundant.

Reply:
76218d756741202c62c33a649d4203fdcf8ecd14